### ISSUE TITLE
Custom subdomain

### ISSUE THREAD

**Issue by pyciko** [digitalocean/supabase-on-do#42]
I'd like to use a different subdomain for supabase, namely api.domain.com. What are the modifications I need to do to achieve that? By quick searching "supabase.", I found (I think) 3 relevant places and changed them to "api.", then I used terraform to build the infrastructure, and unfortunately, that didn't work. I just broke the whole thing:) It didn't respond to any requests either to api. or supabase.

So is there a list of places where I need to change the subdomain to achieve the goal?
Thx in advance

**Comment by timothymamo** [comment]
Hey @pyciko I believe the following are places where you would need to change the subdomain to `api` for the changes to work:

* In the `docker-compose.yml` file on [line 19](https://github.com/digitalocean/supabase-on-do/blob/main/packer/supabase/docker-compose.yml#L19)
* In the `supabase.subdomain.conf` file on [line 5](https://github.com/digitalocean/supabase-on-do/blob/main/packer/supabase/supabase.subdomain.conf#L5).
* In the `.env.tfpl` file on [line 62](https://github.com/digitalocean/supabase-on-do/blob/main/terraform/files/.env.tftpl#L62) and [line 93](https://github.com/digitalocean/supabase-on-do/blob/main/terraform/files/.env.tftpl#L93)
* In the `network.tf` file on [line 25](https://github.com/digitalocean/supabase-on-do/blob/main/terraform/network.tf#L25)

I'll see if I can modify the code to make the subdomain modifiable (although the only file I can see being an issue is the `supabase.subdomain.conf`, might have to make that a template).

In the meantime hope this helps and I managed to get all the places that need to be changed.

**Comment by pyciko** [comment]
Okay, thx, I'll try a bit later.
Another question on the creation flow:
It requires me to specify a website and throws an error if that website is not added to DO domains list. But the problem is, the website is registered via Cloudflare and uses Cloudflare nameservers. DO domains page states that I need to change the nameservers to those provided by DO, but this is 1) expensive (requires a 250$/month) subscription on Cloudflare 2) I'd really like to use Cloudflare for that task anyway. 

To avoid the domain error, I'm adding domain.com to DO domains list, but actually it's like a "dead record" because in reality the CF nameservers are still used. It helps me to suppress this issue, but I'm wondering if such technique could affect the deployment process and in either case, do you have any ideas on how to handle this whole thing elegantly?)

**Comment by timothymamo** [comment]
The following [comment](https://github.com/digitalocean/supabase-on-do/discussions/17#discussioncomment-5171161) should help with domains registered in Cloudflare.

**Comment by pyciko** [comment]
Thanks!

**Comment by pyciko** [comment]
Just a small question: is there any way to separate the admin panel subdomain from api access subdomain?
And also, is it possible to hide the admin panel from internet access is future altogether?

**Comment by timothymamo** [comment]
Hi @pyciko unsure if this is possible, this seems to be a question for the [supabase team](https://github.com/supabase/supabase).

With regards to hiding the admin panel from the internet, not 100% sure what you mean. Disabling internet access of the admin page would also mean disabling it for yourself too.

**Comment by pyciko** [comment]
Yeah I meant if it was possible to move it to another subdomain, then I can (for example) make a CloudFlare rule to only accept connection from certain IPs.

**Comment by timothymamo** [comment]
You can limit access to your subdomain to a specific set of IPs by modifying [line 48](https://github.com/digitalocean/supabase-on-do/blob/main/terraform/network.tf#L48) and [line 54](https://github.com/digitalocean/supabase-on-do/blob/main/terraform/network.tf#L54) in the [network.tf](https://github.com/digitalocean/supabase-on-do/blob/main/terraform/network.tf) file.

If you do this I would also suggest you limit the IPs that can access the VM via [ssh](https://github.com/digitalocean/supabase-on-do/blob/main/terraform/terraform.tfvars.example#L32) and [db port](https://github.com/digitalocean/supabase-on-do/blob/main/terraform/terraform.tfvars.example#L34) in your tfvars file. Or disable access to these ports completely by setting the appropriate booleans [[ssh](https://github.com/digitalocean/supabase-on-do/blob/main/terraform/terraform.tfvars.example#L31) and [db](https://github.com/digitalocean/supabase-on-do/blob/main/terraform/terraform.tfvars.example#L33)] to false.

**Comment by pyciko** [comment]
Sorry, I may misunderstand something, since you said that separation of admin subdomain from main API subdomain is not possible, I guess limiting IPs list is not appropriate because it will also limit them for the main API...

**Comment by timothymamo** [comment]
Unsure if you can or cannot separate the admin and api subdomain, I'd ask in the supabase repo, but yes, if you limit to a specific range of IPs you'll be doing that for the api too (which you might want to do if you're not opening up your supabase to the world and only to a specific site, and maybe your personal IP)

**Comment by pyciko** [comment]
understood

**Comment by pyciko** [comment]
Okay so I went through the process using your suggestions. When doing terraform plan, it still yelled at me that the domain has not been added to DO domains, so I still had to add it there. Everything else went without errors, CF records were created successfully, but...

It is still unresponsive. I waited 15 mins after Droplet creation, and I still can't access it neither on api. nor on supabase.

**Comment by pyciko** [comment]
Some additional info:
1) There are no places with "digitalocean_record" left (checked in VS Code)
2) Droplet, Volume, Space and Reserved IP created successfully. 
3) CF records added successfully.

**Comment by ghost** [comment]
What is the status of the Droplet? Can you ssh into it and run a `docker ps -a`? Can you check the logs for the swag container `docker logs swag`? What results do you get when you run a `dig <your supabase domain>` and a `traceroute <your supabase domain>`?

**Comment by pyciko** [comment]
I guess I see that there's a problem with Let's Encrypt certificate...

**Comment by timothymamo** [comment]
So the error message below is stating that certbot (which requests the certificate from let's encrypt for you) couldn't verify that the domain exists on DigitalOcean.

```
Hint: The Certificate Authority failed to verify the DNS TXT records created by --dns-digitalocean. Ensure the above domains are hosted by this DNS provider, or try increasing --dns-digitalocean-propagation-seconds (currently 10 seconds).
```

If you are using CloudFlare for your DNS than you need to modify the `DNSPLUGIN` environment variable to use `cloudflare` and have the [`digitalocean.ini.tpfl`](https://github.com/digitalocean/supabase-on-do/blob/main/terraform/files/digitalocean.ini.tftpl) to be similar to  the following [file](https://github.com/linuxserver/docker-swag/blob/master/root/defaults/dns-conf/cloudflare.ini) as specified in the swag repo. Something I seem to have missed in the linked comment but I assume the previous person figured it out.

You say that you put the domain on DigitalOcean, what is running in `CloudFlare`? As there are a fare few changes so unsure what the repo looks like now.

**Comment by pyciko** [comment]
Sorry, I choose wrong words:)
I mean I added the domain to DO domains list, but it is still fully operated via CloudFlare. On DO it's just a dummy to suppress the error. The repo has following changes: replaced "supabase" with "api" in 5 places you mentioned, replaced "digitalocean_record" with "cloudflare_record" in all places where it was used (and replaced domain with zone_id since it's the term used in CloudFlare config) + added CloudFlare to providers list with an api token. That's all.

About the main part of your suggestion. After doing all that, do I need to run terraform build and that will redeploy  the changes? Or do I need to do anything else? Sorry for stupid questions, this is my first time working with packer and terraform:)

**Comment by ghost** [comment]
You would have to recreate the image with packer as you are changing the `docker-compose.yml` file and once that is done rerun terraform to redploy the droplets with the new images (terraform picks up the most recent created image).

No worries, there are no stupid questions, it's just a little more difficult to try and debug/help when you're not looking at the same screen :).

**Comment by pyciko** [comment]
Do i need to manually destroy the previously created resources and/or CloudFlare records?

**Comment by ghost** [comment]
No terraform will destroy any resources that are going to be recreated/changed/updated. 

In fact anything you create with terraform you should never touch manually afterwards as you would get [drift](https://developer.hashicorp.com/terraform/tutorials/state/resource-drift) from what is in your terraform state file and what is actually deployed.

**Comment by pyciko** [comment]
Still doesn't work. I reran the commands you asked above, docker ps -a output is the same, dig output is the same, but **traceroute is different:**

>  1  192.168.1.1 (192.168.1.1)  4.263 ms  3.763 ms  3.276 ms
 2  10.203.0.1 (10.203.0.1)  3.748 ms  3.627 ms  3.530 ms
 3  212.111.193.1 (212.111.193.1)  11.352 ms  11.299 ms  11.166 ms
 4  uran-ap1-ias-geant-gw.kie.ua.geant.net (83.97.89.192)  11.302 ms  11.665 ms  11.417 ms
 5  *^C
pyciko@MBA terraform % traceroute api.yalrow.com
traceroute to api.yalrow.com (188.166.193.208), 64 hops max, 52 byte packets
 1  192.168.1.1 (192.168.1.1)  3.704 ms  3.119 ms  3.138 ms
 2  10.203.0.1 (10.203.0.1)  3.559 ms  3.631 ms  3.515 ms
 3  212.111.193.1 (212.111.193.1)  11.229 ms  11.355 ms  12.242 ms
 4  uran-ap1-ias-geant-gw.kie.ua.geant.net (83.97.89.192)  11.365 ms  11.465 ms  11.285 ms
 5  * * *
 6  * * *
 7  ae8.mx1.ams.nl.geant.net (62.40.98.66)  40.918 ms  41.732 ms  41.041 ms
 8  ae40.ams3-edge1.digitalocean.com (80.249.211.163)  42.447 ms  42.607 ms  42.481 ms
 9  * * *
10  * * *
11  * * *
12  * * *
13  * * *
14  * * *

**and swag still has the problem**

> [migrations] started
[migrations] 01-nginx-site-confs-default: executing...
[migrations] 01-nginx-site-confs-default: succeeded
[migrations] done

-------------------------------------
          _         ()
         | |  ___   _    __
         | | / __| | |  /  \
         | | \__ \ | | | () |
         |_| |___/ |_|  \__/


Brought to you by linuxserver.io
-------------------------------------

To support the app dev(s) visit:
Certbot: https://supporters.eff.org/donate/support-work-on-certbot

To support LSIO projects visit:
https://www.linuxserver.io/donate/
-------------------------------------
GID/UID
-------------------------------------

User uid:    1000
User gid:    1000
-------------------------------------

Setting resolver to  127.0.0.11
Setting worker_processes to 1
generating self-signed keys in /config/keys, you can replace these with your own keys if required
............+...........+............+..................+....+.....+..........+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++*......+...+...............+..+...+...+....+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++*.+...+..+............+...+.+......+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
..+.........+..+...+....+.....+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++*..+.....+....+.....+.............+......+......+...+..+...+...+...+.........+.+......+...+...+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++*.+.+...+............+.........+.....+..........+.....+....+..+.......+..+.+.........+.....+................+..+...+...+.+......+.........+..................+..+..........+.........+...........+...+.+.........+.....+.+............+...+.....+......+.+..+............+.........+.+...+........+.............+..+...+......+.+..+...+..................................+....................+...+.+..+...+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
-----
# do daily/weekly/monthly maintenance
# min   hour    day     month   weekday command
*/15    *       *       *       *       run-parts /etc/periodic/15min
0       *       *       *       *       run-parts /etc/periodic/hourly
0       2       *       *       *       run-parts /etc/periodic/daily
0       3       *       *       6       run-parts /etc/periodic/weekly
0       5       1       *       *       run-parts /etc/periodic/monthly
# renew letsencrypt certs
8       2       *       *       *       /app/le-renew.sh >> /config/log/letsencrypt/letsencrypt.log 2>&1
Variables set:
PUID=1000
PGID=1000
TZ=Europe/Berlin
URL=yalrow.com
SUBDOMAINS=api
EXTRA_DOMAINS=
ONLY_SUBDOMAINS=true
VALIDATION=dns
CERTPROVIDER=
DNSPLUGIN=cloudflare
EMAIL=
STAGING=

Created .donoteditthisfile.conf
Using Let's Encrypt as the cert provider
SUBDOMAINS entered, processing
SUBDOMAINS entered, processing
Only subdomains, no URL in cert
Sub-domains processed are:  -d api.yalrow.com
No e-mail address entered or address invalid
dns validation via cloudflare plugin is selected
Generating new certificate
Saving debug log to /var/log/letsencrypt/letsencrypt.log
Account registered.
Requesting a certificate for api.yalrow.com
Unsafe permissions on credentials configuration file: /config/dns-conf/cloudflare.ini
Error determining zone_id: 9103 Unknown X-Auth-Key or X-Auth-Email. Please confirm that you have supplied valid Cloudflare API credentials. (Did you enter the correct email address and Global key?)
Ask for help or search for solutions at https://community.letsencrypt.org. See the logfile /var/log/letsencrypt/letsencrypt.log or re-run Certbot with -v for more details.
ERROR: Cert does not exist! Please see the validation error above. Make sure you entered correct credentials into the /config/dns-conf/cloudflare.ini file.

If it can help somehow, I'm using Full(strict) SSL/TLS encryption mode on CF and I already have a certificate for domain.com and *.domain.com that is valid until 2038. Maybe I can use it somehow...

**Comment by ghost** [comment]
Seems like the credentials in the `cloudflare.ini` file are not correct.

So to change the `ini` file you'll have to do a few things:
Create a new file in the [files](https://github.com/digitalocean/supabase-on-do/tree/main/terraform/files) section called `cloudflare.ini.tftpl` with the following:
```
dns_cloudflare_email = ${TF_CLOUDFLARE_EMAIL}
dns_cloudflare_api_key = ${TF_CLOUDFLARE_API_KEY}
```

In the locals.tf file remove line 136 and add the following:

```
cloudflare_ini = templatefile("${path.module}/files/cloudflare.ini.tftpl",
    {
      TF_CLOUDFLARE_EMAIL = "${var.cloudflare_email}",
      TF_CLOUDFLARE_API_KEY = "${var.cloudflare_api_key}"
    }
  )
```

and change line 164 and 168 in the `locals.tf` file to have the following:
```
path        = "/root/supabase/cloudflare.ini"
...
content     = base64encode("${local.cloudflare_ini}")
```

In your `docker-compose.yml` file change line 24 to the following:
```
      - ./cloudflare.ini:/config/dns-conf/cloudflare.ini
```

I think that is all.

Trying to add your own would need changes in the conf file and understanding of nginx (which might be even more cumbersome to do currently).

**Comment by pyciko** [comment]
It worked!
Huge thanks!
I will share my modifications a bit later once I remove the API keys and stuff..
Also, I guess I might've exposed some sensitive info in the logs I sent earlier, can I remove logs from these messages?

**Comment by ghost** [comment]
Awesome! Glad it worked.

You should be able to edit your comments and remove any info which is sensitive.

It would be great if you could share your setup in the Discussion section for anyone who needs it in the future.

**Comment by pyciko** [comment]
Done, created a discussion with the instructions.
Another question: now I want to integrate external auth providers like I did in supabase-provided version. I found instructions on how set things up in .env file, but I'm confused with terraform setup of this. Theres an "environment" section in docker-compose.yml, and then there's .env.tftpl file in terraform/files. They seems to be duplicating each other. So what do I need to modify?
